### DISCUSSION TITLE
handling selectors with '.' within the element name

### DISCUSSION THREAD

**Discussion by link4531** [TomWright/dasel#139] [Q&A] [answered]
Edit It may be more appropriate to refer to as 'node name' or 'selector value'
say I have a yaml/json structure such as 

```>code
>this.file:
>   -this.attribute: value
```

what syntax do I need to provide dasel to updating this.attribute to 'newvalue'
```>code
>./dasel -p yaml -s this.\file.this.\attribute newvalue
```
this command breaks out ths,file,this,attribute in to separate nodes. I was hoping the '\\' character would solve this but it didn't help. Any suggestions?

**Comment by TomWright** [comment, marked as answer]
You just need to escape the dot. The backslash does do this, but you are putting in the wrong place. The backslash needs to come *before* the character you want to escape.

E.g.
```bash
 echo 'this.file:
  - this.attribute: value' | dasel put string -p yaml '.this\.file.[0].this\.attribute' newvalue
```

```yaml
this.file:
- this.attribute: newvalue
```

**Reply by link4531** [reply]
(ノ°Д°）ノ︵ ┻━┻

   * deep sigh * 

Its lessons like this, that will give me night terrors 10 years down the road.

Thank you sir for assisting me.

I swear I had tried every method of using '\\' even the example you provided ( obviously I was wrong ) 
What I did discover is every way to do it wrong...

**Reply by postscripter8** [reply]
Hello Sir, I just found this fantastic tool as I attempted to automate containerd configuration process for my lab. Unfortunately the latest dasel version tokenizes queries differently and I can't quite get it to work with my config.toml file. The example you provided fails to parse with

> dasel: error: error parsing selector: failed to tokenize: unexpected token: \ at position 5.

I need to adjust SystemdCgroup value in the minimized example below. I don't know whose vile idea it was to put apostrophes in the key name `[plugins.'io.containerd.cri.v1.runtime'.containerd.runtimes.runc.options]`

```
version = 4
[plugins]
  [plugins.'io.containerd.cri.v1.runtime']
    enable_selinux = false

    [plugins.'io.containerd.cri.v1.runtime'.containerd]
      default_runtime_name = 'runc'

      [plugins.'io.containerd.cri.v1.runtime'.containerd.runtimes]
        [plugins.'io.containerd.cri.v1.runtime'.containerd.runtimes.runc]
          runtime_type = 'io.containerd.runc.v2'

          [plugins.'io.containerd.cri.v1.runtime'.containerd.runtimes.runc.options]
            ShimCgroup = ''
            SystemdCgroup = false
```

I attempted query string `'plugins.io\.containerd\.cri\.v1\.runtime.containerd.runtimes.runc.options'` but it failed in a similar manner.

> dasel: error: error parsing selector: failed to tokenize: unexpected token: \ at position 10.

**Reply by postscripter8** [reply]
Oh, I just found my exact use case in https://github.com/TomWright/dasel/discussions/130.

**Reply by postscripter8** [reply]
Eh... I got bored waiting for a reply and decided to stop being lazy for once. Found the answer in https://github.com/TomWright/dasel/blob/master/selector/lexer/tokenize.go#L185

This does what I need in dasel v3.11.2:
```
<config.toml dasel -i toml --root "plugins['io.containerd.cri.v1.runtime'].containerd.runtimes.runc.options.SystemdCgroup=true"
```
